File: advent_php/advent_five.php
<?php

function parse_input($input) {
    list($rules_section, $updates_section) = explode("\n\n", trim($input));
    $rules = array_map('trim', explode("\n", $rules_section));
    $updates = array_map('trim', explode("\n", $updates_section));
    return [$rules, $updates];
}

function build_graph($rules) {
    $graph = [];
    foreach ($rules as $rule) {
        list($before, $after) = explode('|', $rule);
        if (!isset($graph[$before])) {
            $graph[$before] = [];
        }
        $graph[$before][] = $after;
    }
    return $graph;
}

function is_valid_update($update, $graph) {
    $pages = explode(',', $update);
    $positions = array_flip($pages);
    foreach ($graph as $before => $afters) {
        if (isset($positions[$before])) {
            foreach ($afters as $after) {
                if (isset($positions[$after]) && $positions[$before] > $positions[$after]) {
                    return false;
                }
            }
        }
    }
    return true;
}

function find_middle_page($update) {
    $pages = explode(',', $update);
    $middle_index = floor(count($pages) / 2);
    return $pages[$middle_index];
}

function sum_middle_pages($input) {
    list($rules, $updates) = parse_input($input);
    $graph = build_graph($rules);
    $sum = 0;
    foreach ($updates as $update) {
        if (is_valid_update($update, $graph)) {
            $sum += find_middle_page($update);
        }
    }
    return $sum;
}

$input = file_get_contents('input.txt');

echo "Sum of middle pages: " . sum_middle_pages($input) . "\n";

?>

File: advent_py/advent_five.py
def parse_input(input_data):
    rules_section, updates_section = input_data.strip().split("\n\n")
    rules = [line.strip() for line in rules_section.split("\n")]
    updates = [line.strip() for line in updates_section.split("\n")]
    return rules, updates

def build_graph(rules):
    graph = {}
    for rule in rules:
        before, after = rule.split('|')
        if before not in graph:
            graph[before] = []
        graph[before].append(after)
    return graph

def is_valid_update(update, graph):
    pages = update.split(',')
    positions = {page: idx for idx, page in enumerate(pages)}
    for before, afters in graph.items():
        if before in positions:
            for after in afters:
                if after in positions and positions[before] > positions[after]:
                    return False
    return True

def find_middle_page(update):
    pages = update.split(',')
    middle_index = len(pages) // 2
    return int(pages[middle_index])

def sum_middle_pages(input_data):
    rules, updates = parse_input(input_data)
    graph = build_graph(rules)
    total_sum = 0
    for update in updates:
        if is_valid_update(update, graph):
            total_sum += find_middle_page(update)
    return total_sum

with open("input.txt", "r") as file:
    input_data = file.read()

print("Sum of middle pages:", sum_middle_pages(input_data))